compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/NodeFinder.java

```diff
@@ -0,0 +1,1484 @@
+import org.wso2.ballerinalang.compiler.tree.statements.BLangWhile;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangWorkerSend;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangXMLNSStatement;
+import org.wso2.ballerinalang.compiler.tree.types.BLangArrayType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangConstrainedType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangErrorType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangFiniteTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangFunctionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangIntersectionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangLetVariable;
+import org.wso2.ballerinalang.compiler.tree.types.BLangObjectTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangRecordTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangStreamType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangTupleTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangUnionTypeNode;
+import org.wso2.ballerinalang.compiler.tree.types.BLangUserDefinedType;
+import org.wso2.ballerinalang.compiler.tree.types.BLangValueType;
+import org.wso2.ballerinalang.compiler.util.diagnotic.DiagnosticPos;
+
+import java.util.List;
+
+import static org.ballerinalang.model.symbols.SymbolOrigin.VIRTUAL;
+
+/**
+ * Finds the enclosing AST node for the given position.
+ *
+ * @since 2.0.0
+ */
+class NodeFinder extends BLangNodeVisitor {
+
+    private LineRange range;
+    private BLangNode enclosingNode;
+
+    BLangNode lookup(BLangCompilationUnit unit, LineRange range) {
+        this.range = range;
+        this.enclosingNode = null;
+
+        for (TopLevelNode node : unit.topLevelNodes) {
+            if (!PositionUtil.withinRange(this.range, node.getPosition()) || isLambdaFunction(node)) {
+                continue;
+            }
+
+            ((BLangNode) node).accept(this);
+
+            if (this.enclosingNode != null) {
+                break;
+            }
+        }
+
+        return this.enclosingNode;
+    }
+
+    private void lookupNodes(List<? extends BLangNode> nodes) {
+        for (BLangNode node : nodes) {
+            if (!PositionUtil.withinRange(this.range, node.pos)) {
+                continue;
+            }
+
+            node.accept(this);
+
+            if (this.enclosingNode == null && !node.internal && setEnclosingNode(node, node.pos)) {
+                return;
+            }
+        }
+    }
+
+    private void lookupNode(BLangNode node) {
+        if (node == null) {
+            return;
+        }
+
+        if (PositionUtil.withinRange(this.range, node.pos)) {
+            node.accept(this);
+        }
+
+        if (this.enclosingNode == null && !node.internal) {
```

Comment:
Do we need this null check since we do the null check within the `setEnclosingNode` method ?

Reply:
Modified the logic related to this a bit. The `setEnclosingNode()` is usually used for nodes with names (e.g., variables, constants), to see if the provided range is the range of the name of it. With this latest cleanup, this is now required since now, for the typical case, this does what `setEnclosingNode()` method does.